packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -229,5 +233,50 @@ export class EpochTracker {
     }
 }
 
-export type FetchType = "blob" | "createBlob" | "createFile" | "joinSession" | "ops" | "other" | "snapshotTree" |
-    "treesLatest" | "uploadSummary" | "push";
+export class EpochTrackerWithRedemption extends EpochTracker {
+    private readonly treesLatestDeferral = new Deferred<void>();
+
+    protected validateEpochFromResponse(
+        epochFromResponse: string | undefined | null,
+        fetchType: FetchType,
+        fromCache: boolean = false,
+    ) {
+        super.validateEpochFromResponse(epochFromResponse, fetchType, fromCache);
+
+        // Any successful call means we have access to a file, i.e. any redemption that was required already happened.
+        // That covers cases of "treesLatest" as well as "getVersions" or "createFile" - all the ways we can start
+        // exploring a file.
+        this.treesLatestDeferral.resolve();
+    }
+
+    public async fetchAndParseAsJSON<T>(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        fetchType: FetchType,
+        addInBody: boolean = false,
+    ): Promise<IOdspResponse<T>> {
+        // Optimize the flow if we know that treesLatestDeferral was already completed by the timer we started
+        // joinSession call. If we did - there is no reason to repeat the call as it will fail with same error.
+        const completed = this.treesLatestDeferral.isCompleted;
+
+        try {
+            return await super.fetchAndParseAsJSON<T>(url, fetchOptions, fetchType, addInBody);
+        } catch (error) {
+            // Only handling here treesLatest. If createFile failed, we should never try to do joinSession.
+            // Similar, if getVersions failed, we should not do any further storage calls.
+            // So treesLatest is the only call that can have parallel joinSession request.
+            if (fetchType === "treesLatest") {
+                this.treesLatestDeferral.reject(error);
+            }
+            if (fetchType !== "joinSession" || error.statusCode !== 404 || !completed) {
```

Comment:
```suggestion
            if (fetchType !== "joinSession" || error.statusCode !== 404 || completed) {
```
Switching this to completed to throw the error if we've already attempted a retry in the past.
This successfully initiates the retry in the 3p prototype scenario